src/main/scala/mesosphere/marathon/raml/VolumeConversion.scala

```diff
@@ -33,106 +45,140 @@ trait VolumeConversion extends ConstraintConversion with DefaultConversions {
+      case DiskType.Root => PersistentVolumeType.Root
+    })
+    PersistentVolumeInfo(`type` = pvType, size = pv.size, maxSize = pv.maxSize, profileName = pv.profileName,
+      constraints = pv.constraints.toRaml[Set[Seq[String]]])
+  }
+
+  implicit val volumeWrites: Writes[state.VolumeWithMount, AppVolume] = Writes { volumeWithMount =>
+
+    implicit val externalVolumeWrites: Writes[state.ExternalVolumeInfo, ExternalVolumeInfo] = Writes { ev =>
+      ExternalVolumeInfo(size = ev.size, name = Some(ev.name), provider = Some(ev.provider), options = ev.options)
     }
 
+    val volume = volumeWithMount.volume
+    val mount = volumeWithMount.mount
     volume match {
-      case dv: state.DockerVolume => AppDockerVolume(
-        volume.containerPath,
-        dv.hostPath,
-        mode = volume.mode.toRaml)
+      case dv: state.HostVolume => AppHostVolume(
+        containerPath = mount.mountPath,
+        hostPath = dv.hostPath,
+        mode = mount.readOnly.toRaml)
       case ev: state.ExternalVolume => AppExternalVolume(
-        volume.containerPath,
+        containerPath = mount.mountPath,
         external = ev.external.toRaml,
-        mode = volume.mode.toRaml)
+        mode = mount.readOnly.toRaml)
       case pv: state.PersistentVolume => AppPersistentVolume(
-        volume.containerPath,
+        containerPath = mount.mountPath,
         persistent = pv.persistent.toRaml,
-        mode = volume.mode.toRaml)
+        mode = mount.readOnly.toRaml)
       case sv: state.SecretVolume => AppSecretVolume(
-        volume.containerPath,
+        containerPath = mount.mountPath,
         secret = sv.secret
       )
     }
   }
 
-  implicit val volumeReads: Reads[AppVolume, state.Volume] = Reads {
+  implicit val volumeReads: Reads[AppVolume, state.VolumeWithMount] = Reads {
     case v: AppExternalVolume => volumeExternalReads.read(v)
     case v: AppPersistentVolume => volumePersistentReads.read(v)
-    case v: AppDockerVolume => volumeDockerReads.read(v)
+    case v: AppHostVolume => volumeHostReads.read(v)
     case v: AppSecretVolume => volumeSecretReads.read(v)
     case unsupported => throw SerializationFailedException(s"unsupported app volume type $unsupported")
   }
 
-  implicit val volumeExternalReads: Reads[AppExternalVolume, state.Volume] = Reads { vol =>
-    val info = ExternalVolumeInfo(
-      size = vol.external.size,
-      name = vol.external.name.getOrElse(throw SerializationFailedException("external volume requires a name")),
-      provider = vol.external.provider.getOrElse(throw SerializationFailedException("external volume requires a provider")),
-      options = vol.external.options
+  implicit val volumeExternalReads: Reads[AppExternalVolume, state.VolumeWithMount] = Reads { volumeRaml =>
+    val info = state.ExternalVolumeInfo(
+      size = volumeRaml.external.size,
+      name = volumeRaml.external.name.getOrElse(
+        throw SerializationFailedException("external volume requires a name")),
```

Comment:
shouldn't this be handled through validation layer?

also the indentation seems weird (meaning a bit different) to me...

Reply:
It is. But since `scapegoat` complains about `.get` calls. Also, it is better to double-check than to be sorry :)

The indentation looks fine to me. What do you dislike about it?

Reply:
indentation should just be handled by our formatter :)

I think I'd like to make it clear that this error message should never be seen. Maybe throw an IllegalStateException?

Reply:
```
/**
  * Thrown for errors during [[Normalization]]. Validation should normally be checking for invalid data structures
  * that would lead to errors during normalization, so these exceptions are very unexpected.
  * @param msg provides details
  */
case class NormalizationException(msg: String) extends Exception(msg)

case class SerializationFailedException(message: String) extends Exception(message)
```

These exceptions are said to be very unexpected, whatever it means :)

Reply:
I remember @jdef and I having this same conversation :)